### PR TITLE
QskSpinBox

### PR DESCRIPTION
**Pull request by rick-vogel**
Solves #105

image

- [ ] implement focus traversal
- [ ] improve size hints & policies
- [ ] apply clang-format
- [ ] fix focus state pre subcontrol
- [ ] BUG: increment to 0.3, decrement 3x returns value less than 0
- [ ] style for squiek skin
- [ ] allow symbols / images for button icons instead of +/-

### CODE REVIEW

File: examples/gallery/main.cpp [hunk 1]
```diff
@@ -199,6 +200,9 @@ namespace
             tabView->addTab( "Text\nInputs", new TextInputPage() );
             tabView->addTab( "Selectors", new SelectorPage() );
             tabView->addTab( "Dialogs", new DialogPage() );
+            tabView->addTab( "SpinBoxes", new SpinBoxPage() );
+
+            tabView->setCurrentIndex(tabView->count() - 1);
```
Comment (rick-vogel) [inline examples/gallery/main.cpp:205]: remove before merge

File: examples/gallery/spinbox/SpinBoxPage.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  };
+
+  auto* const grid = new QskGridBox(this);
+  constexpr int cols = 5;
+  for(const auto& layout : layouts.keys())
+  {
+    const auto x = grid->elementCount() % cols;
+    const auto y = grid->elementCount() / cols;
+    auto* const column = new QskLinearBox(Qt::Vertical, grid);
+    auto* const label = new QskTextLabel(layouts.value(layout), column); // TODO put label either on top or on the bottom
```
Comment (rick-vogel) [inline examples/gallery/spinbox/SpinBoxPage.cpp:43]: Place the label right on top or at the bottom

File: skins/material3/QskMaterial3Skin.cpp [hunk 3]
```diff
@@ -672,6 +675,76 @@ void Editor::setupSlider()
...
+  setPadding( Q::TextPanel, 5_dp );
+  setBoxShape( Q::TextPanel, 4_dp, 4_dp, 0, 0 );
+  setBoxBorderMetrics( Q::TextPanel, 0, 0, 0, 1_dp );
+  setBoxBorderColors( Q::TextPanel, m_pal.onSurface );
+
+  setBoxBorderMetrics( Q::TextPanel | Q::Focused, 0, 0, 0, 2_dp );
+  setBoxBorderColors( Q::TextPanel | Q::Focused, m_pal.primary );
+
+//  setBoxBorderMetrics( Q::TextPanel | Q::Editing, 0, 0, 0, 2_dp );
+//  setBoxBorderColors( Q::TextPanel | Q::Editing, m_pal.primary );
```
Comment (rick-vogel) [inline skins/material3/QskMaterial3Skin.cpp:733]: Currently no editing

File: src/controls/QskSpinBox.cpp [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+    using LUT = std::array<FocusIndeces,4>;
+    if(layout == Qt::AlignLeft) return LUT{Inc,Dec,None,Text}[m_focusIndex];
+    if(layout == Qt::AlignRight) return LUT{Inc,None,Text,Dec}[m_focusIndex];
+    if(layout == Qt::AlignHCenter) return LUT{Text,Inc,None,Dec}[m_focusIndex];
+    if(layout == Qt::AlignTop) return LUT{Inc,Dec,None,Text}[m_focusIndex];
+    if(layout == Qt::AlignBottom) return LUT{Inc,None,Text,Dec}[m_focusIndex];
+    if(layout == Qt::AlignVCenter) return LUT{None,Dec,Text,Inc}[m_focusIndex];
+    if(layout == (Qt::AlignLeft | Qt::AlignVCenter)) return LUT{None,Inc,Dec,Text}[m_focusIndex];
+    if(layout == (Qt::AlignRight | Qt::AlignVCenter)) return LUT{Text,None,Dec,Inc}[m_focusIndex];
+
```
Comment (rick-vogel) [inline src/controls/QskSpinBox.cpp:87]: TODO Complete

File: src/controls/QskSpinBox.cpp [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+    using LUT = std::array<FocusIndeces,4>;
+    if(layout == Qt::AlignLeft) return LUT{None,Dec,Text,Inc}[m_focusIndex];
+    if(layout == Qt::AlignRight) return LUT{None,Inc,Dec,Text}[m_focusIndex];
+    if(layout == Qt::AlignHCenter) return LUT{None,Dec,Text,Inc}[m_focusIndex];
+    if(layout == Qt::AlignTop) return LUT{Text,None,Dec,Inc}[m_focusIndex];
+    if(layout == Qt::AlignBottom) return LUT{None,Inc,Dec,Text}[m_focusIndex];
+    if(layout == Qt::AlignVCenter) return LUT{Text,Inc,None,Dec}[m_focusIndex];
+    if(layout == (Qt::AlignLeft | Qt::AlignVCenter)) return LUT{Inc,None,Text,Dec}[m_focusIndex];
+    if(layout == (Qt::AlignRight | Qt::AlignVCenter)) return LUT{Inc,Dec,None,Text}[m_focusIndex];
+
```
Comment (rick-vogel) [inline src/controls/QskSpinBox.cpp:106]: TODO Complete

**Comment by uwerat** [comment]
As the focus chain usually depends on the geometries of the up/down and text subcontrols it might be worth to check if it is not possible to find the chain from the subcontrolRects - instead of having these tables, that depend on the alignmentHint.

**Pull request merged**